### ISSUE TITLE
Add support for polyphonic generation

### ISSUE THREAD

**Issue by eraoul** [tensorflow/magenta#123]
I would guess that several people are working on this already; I'm on it as well.

**Comment by prafiles** [comment]
I'd like to contribute to the code as well, if that's possible.

**Comment by prafiles** [comment]
Hey guys sorry for bothering everyone here again. I'd like to ask why are we ignoring the polyphonic music while training? is there a specific reason (as in we should train it in something we aren't able to generate back yet?)

Because I looked around the code a bit to find a flag whose default is set to ignore polyphonic music. I have this amazing collection of polyphonic music that I'd like to train on, just for a bit of fun.

**Comment by douglaseck** [comment]
Magenta intern @kastnerkyle has done some nice polyphonic work. But it's not ready for release.  Messy code, Kyle? :-) In general the reason why is that it's a lot harder than monophonic prediction. With monophonic prediction, you can sample from a single softmax, but once you add multiple notes (and worse, maybe 1,2,3,4,5 .. however many at one time) the possible combinations of what could happen at any timestep explodes.  One answer is to do conditional prediction. For example, hard limit to K voices and predict them conditionally (voice 1 is predicted, then voice 2 conditional on voice 1, . . . and so forth).  Or even predict the number of notes at a given timestep, then predict them.  I don't think there's a single right way to do this and I hope we can make it easy (read "not horribly painful") to try different things.  Really these sorts of decisions are the bread-and-butter of machine learning. What are you modeling? Where are the assumptions? The "right" answer is probably always to model the joint distribution of everything. But often that's intractable, and making it tractable by introducing factorizations like conditional prediction, is part of the fun of ML.

**Comment by prafiles** [comment]
Thanks so much for sharing the information. This explains a lot. 

One question however, what did you mean by `(read "not horribly painful") `. Although I think I understand, what it meant to convey was that the problem at hand wasn't too easy, is there some documentation that I missed out completely?

**Comment by kastnerkyle** [comment]
Usually to solve these kinds of fundamental problems, you are working at
the level of Tensorflow itself. I cannot think of any easy way to get the
flexibility (at least I) needed for this task without dropping down to that
level. Hopefully the rest of my internship (til September 2.) will focus on
getting some code cleanup and trying to integrate with the existing Magenta
API. The model itself is not terribly complicated, but there are some
tricky things data-wise I think, especially handling other people's files
and getting different voices out.

Plus, my code still has little scars and stubs from all my failed
ideas/models :p It's not messy exactly, more like wounded in (research)
battle.

**Comment by douglaseck** [comment]
No you didn't miss any documentation. It was idiomatic English, and you
understood it completely. :-) It's not easy to do polyphonic prediction,
but with some care perhaps we can make it happen without too much
refactoring on everyone's part.

**Comment by prafiles** [comment]
Thanks for sharing all this!

Is there someway I personally can contribute to as well? Some tasks still to be done? Some guidelines etc? I am trying to get started with some small contributions to open source.
